test/distributed/_sharded_tensor/test_megatron_prototype.py

```diff
@@ -35,72 +44,165 @@
 
 
 class TestShardedTensorMegatronLinear(ShardedTensorTestBase):
+    class SimpleMegatronLM(torch.nn.Module):
+        def __init__(self, linear_size, rank=None):
+            super().__init__()
+            self.fc1 = torch.nn.Linear(*linear_size[0])
+            self.gelu = torch.nn.GELU()
+            self.fc2 = torch.nn.Linear(*linear_size[1])
+            if rank:
+                self.fc1.cuda(rank)
+                self.fc2.cuda(rank)
+
+        def forward(self, inp):
+            return self.fc2(self.gelu(self.fc1(inp)))
+
+        def weight_override(self, module):
+            self.fc1.weight = clone_module_parameter(module.fc1, "weight")
+            self.fc1.bias = clone_module_parameter(module.fc1, "bias")
+            self.fc2.weight = clone_module_parameter(module.fc2, "weight")
+            self.fc2.bias = clone_module_parameter(module.fc2, "bias")
+
+        def shard_parameter(self, spec):
+            shard_parameter(self.fc1, "weight", spec[0])
+            shard_parameter(self.fc2, "weight", spec[1])
+
+        def get_weight_grad(self):
+            return (self.fc1.weight.grad, self.fc2.weight.grad)
+
+        def get_bias_grad(self):
+            return (self.fc1.bias.grad, self.fc2.bias.grad)
+
+        def get_weights(self):
+            return (self.fc1.weight, self.fc2.weight)
+
+        def get_bias(self):
+            return (self.fc1.bias, self.fc2.bias)
+
+        def get_weight_local_shard(self):
+            return (
+                self.fc1.weight.local_shards()[0].tensor,
+                self.fc2.weight.local_shards()[0].tensor,
+            )
```

Comment:
All of this should not be part of the model. Let us put these helper methods elsewhere. In general the philosophy is that the model should only contain modeling code and nothing else. For example, someone writing a Transformer model should not worry about sharding, weight override etc. Our sharding APIs should provide all this.

Reply:
Ahh, I see. Will do.